Add support for mid-top, mid-bottom, mid-left, mid-right character options

kleinfreund opened the pull request:
This PR adds support for configuring the mid-top, mid-bottom, mid-left and mid-right characters of a table. This solves the following issue:

```
╔═══════════════════╤════╗
║ greetings         │    ║
╟───────╤───────────┼────╢
║ hello │ world     │ :) ║
║       ╟───────────╢    ║
║       │ beautiful │    ║
╟───────╧───────────┼────╢
║ greetings         │    ║
╚═══════════════════╧════╝
```

The second row is surrounded by rows that span across two columns. With custom characters, one could not change the `╤` and `╧` character individually. Similarly, the center cells can now close with a custom characters to columns spanning across rows.

The following table layout is now possible.

```
╔═══════════════════╤════╗
║ greetings         │    ║
╟───────┬───────────┼────╢
║ hello │ world     │ :) ║
║       ├───────────┤    ║
║       │ beautiful │    ║
╟───────┴───────────┼────╢
║ greetings         │    ║
╚═══════════════════╧════╝
```

I tried updating/extending all references to related configuration options. I did not add any tests as I’m not quite familar with writing them how it has been done.

Please let me know if further adjustments need to be made. I’d be happy updating the PR accordingly.

---

The current naming scheme is a bit inconsistent. It doesn’t support describing all character options accurately.

The name `left-mid` refers to `╟` in the example above, while `mid-left` refers to `├`. I understand that changes to existing names are not possible without breaking things. Maybe, in the future, a better naming scheme can be used. It might make use of an additional qualifier specifying whether the characters is for an outer (e.g. `╟`) or inner cell (e.g. `├`).

jamestalmage commented:
That naming is too confusing considering the existing names. Let's come up with something better.

I am open to suggestions, but what about `internalUp: '┴', internalDown: '┬', internalLeft....`? I don't love that either, so please brainstorm something better if you can.

jamestalmage commented:
Also, this will definitely need tests.

kleinfreund commented:
> That naming is too confusing considering the existing names. Let's come up with something better.

Agreed. I propose that this PR only adds new names (likely sticking to a new naming scheme) without chaning older names. This way, users get the new capabilites of specifying inner table characters better. Changing the complete naming scheme can be done in another PR which then would break things.

As we discovered, we need to distinct between outer and inner characters. That should be the first element in the string. The next elements would be row and column. That leaves us with the following naming changes—backwards-compatible and completely reworked:

| Before this PR   | After this PR    | Future               |
|:-----------------|:-----------------|:---------------------|
| top `─`          | top `─`          | outer-top-cell `─`   |
| top-mid `┬`      | top-mid `┬`      | outer-top-mid `┬`    |
| top-left `┌`     | top-left `┌`     | outer-top-left `┌`   |
| top-right `┐`    | top-right `┐`    | outer-top-right `┐`  |
| bottom `─`       | bottom `─`       | outer-bot-cell `─`   |
| bottom-mid `┴`   | bottom-mid `┴`   | outer-bot-mid `┴`    |
| bottom-left `└`  | bottom-left `└`  | outer-bot-left `└`   |
| bottom-right `┘` | bottom-right `┘` | outer-bot-right `┘`  |
| left `│`         | left `│`         | outer-cell-left `│`  |
| left-mid `├`     | left-mid `├`     | outer-mid-left `├`   |
| mid `─`          | mid `─`          | inner-mid-cell `─`   |
| mid-mid `┼`      | mid-mid `┼`      | inner-mid-mid `┼`    |
| right `│`        | right `│`        | inner-cell-right `│` |
| right-mid `┤`    | right-mid `┤`    | inner-mid-right `┤`  |
| middle `│`       | middle `│`       | inner-cell-mid `│`   |
|                  | mid-top `┬`      | inner-top-mid `┬`    |
|                  | mid-bottom `┴`   | inner-bot-mid `┴`    |
|                  | mid-left `├`     | outer-mid-left `├`   |
|                  | mid-right `┤`    | outer-mid-right `┤`  |

The future naming scheme can be formalized like this:

```
chars = Object<character-id, character>
<character-id> = <outer-or-inner>-<row-id>-<column-id>
<outer-or-inner> = 'outer'|'inner'
<row-id> = 'top'|'mid'|'bot'|'cell'
<column-id> = 'left'|'mid'|'right'|'cell'
<character> = String
```

The proposed future naming scheme would also allow to omit one of the prefixes (i.e. leave out `outer-`, implying that it’s the default).

---

As for tests, I did not know how to write them. I’ll try and have a look at it.

I think the tests would be more precise when we use different characters for outer and inner cells. What do you think? That way, we can expect a difference between `╟` and `├`.

kleinfreund commented:
Looking at the tests, I find that using the same characters for the added character names (e.g. outer-mid-left `├` is the same as inner-mid-left `├`) implicitly tests the new behavior since the algorithm in `src/cells.js` was updated. If the algorithm was not correct, the tests would already fail.

What do you think about changing the default character set in order to distinct outer and inner characters? Unfortunately, that would be a breaking change, as users’ tables would suddenly look different.

Alternativel, only the added character names get *different* default characters than their outer counterparts. Doing this would require updating the examples (e.g. `examples/col-and-row-span-examples.js`).

kleinfreund commented:
I added some tests and rebased since a few pull requests were merged inbetween.

I noticed that the following produces incorrect results. The `quuz` column should span across row 2, column 2-3, but it always is treated as if no colSpan was set. This is unrelated to this PR, it breaks on current master as well.

```js
const Table = require('./index.js');
const table = new Table();

table.push(
  [{ content: 'bar', rowSpan: 2 }, 'baz', 'frob']
  , [{ content: 'quuz', colspan: 2 }]
);

console.log(table.toString());
```

jamestalmage commented on test/cell-test.js:
```diff
@@ -583,17 +611,17 @@ describe('Cell',function(){
       it('will draw the top left corner when x=0,y=0',function(){
         cell.x = cell.y = 0;
-        expect(cell.draw('top')).to.equal('┌───────');
+        expect(cell.draw('top')).to.equal('╔═══════');
```
You shouldn't change existing tests unless you are introducing a breaking change. That defeats the purpose as we lose confidence that this change isn't creating a regression for existing users.

kleinfreund replied:
How do the changes made to `test/cell-test.js` allow introducing a regression?

As far as I can tell, the tests in that file behave exactly as before with the addition of testing the additional features of this PR. Namely, configuring different outer and inner characters.

kleinfreund replied:
@jamestalmage Should I revert the changes to the existing tests?

As stated before, I believe the changes to the tests don’t change them in their intended behavior. They only add further coverage of the features that were added.

jamestalmage commented:
> What do you think about changing the default character set in order to distinct outer and inner characters? Unfortunately, that would be a breaking change, as users’ tables would suddenly look different.

I want to avoid that if at all possible. This module has too many dependents, including `npm` itself.

jamestalmage commented:
Nah, let's make the move to the `inner` and `outer` naming convention now. We just need to make sure that the old behavior remains.

The best way to do this would be to create a `normalizeChars` function that will take both the old naming convention, and the new, and output a normalized config that we use internally.

It would look something like:

```js
function normalizeChars(charConfig = {}) {
  const defaultConfig = defaultChars();
  return {
    outerTopCell: charConfig['outer-top-cell'] || charConfig.outerTopCell || charConfig.top || defaultChars.outerTopCell,
     ...
  }
}
```

jamestalmage commented:
What about this structure:

```
chars: {
  outer: {
    top-left: ...
  },
  inner: {
    top-left: ...
  },
  top-left: ...
}
```

When checking for the `top-left` of an **outer** cell, we would first check for `chars.outer.top-left` or followed by `chars.top-left`. For an **inner** cell we would obviously check`chars.inner` instead of `chars.outer`. This allows you to specify characters you want to reuse for both inner and outer cells in the top level `chars.whatever`. It should help keep things more backwards compatible too.

kleinfreund commented:
@jamestalmage I’d like to know where further tests need to be added.

Then, when the tests are ready, I think it would be better to merge this first and rework the naming scheme in another PR.
